### DISCUSSION TITLE
⚠️ Python 3.10

### DISCUSSION THREAD

**Discussion by polm** [explosion/spaCy#9418] [News & Announcements]
Python 3.10 was recently released. Because this is a totally new Python version, spaCy and many of its dependencies need to be recompiled in order to support it. This will take a little while, and based on past experience wheels will probably not be available before November. 

Until wheels are available, we recommend using 3.9 or other supported versions of Python. 

We'll reply to this thread and make other announcements when 3.10 wheels are available.

**Comment by gethinlw** [comment]
I've noticed that I can install Spacy with Python 3.10 (on the current _python:slim_ Docker image, that's based on Python 3.10) if I install _build-essential_ first. I'm guessing that it's compiling it locally on my machine as part of the install. Is there any reason not to do this? I can't say I know enough about pip (or spacy) to understand fully what it's doing.

**Reply by polm** [reply]
There's no strong reason not to do that, no. You might miss out on some CPU optimizations or something but I wouldn't expect it to affect correctness in any way. 

The reason we have this thread is that many users may have trouble setting up a build environment or may not be able to use an isolated Docker environment. This is especially true when GPU dependencies get involved. But if you can build without issue then that's fine too.

**Reply by gethinlw** [reply]
Great, thanks for the impressively quick reply! I'm currently re-running _pip install_ with -vvv to get more feedback, and it looks like it's compiling some of the dependencies, and I'm guessing it'll do the same with Spacy when it's done.

Incidentally, is there any metadata I can check (like [requires-python](https://packaging.python.org/specifications/core-metadata/#requires-python)) before attempting install to pick up an incompatibility? Currently the only way I'm finding out is when my runner times out after an hour (pip isn't failing gracefully, it just gets stuck in a retry loop).

**Reply by polm** [reply]
I think it's very rare for libraries to be proactively incompatible with 3.10, rather than just omitting `requires-python` due to lack of testing, so I think you just have to try building and see what happens unfortunately.

**Reply by gethinlw** [reply]
Ok makes sense, thanks. I've hacked together a workaround using _timeout_ to pick up if there are issues in the future so I'll stick with that for now.

**Reply by adrianeboyd** [reply]
In particular, pay attention to how `numpy` is being built. If you're building it without a BLAS library, the speed will be affected.

There's no particular benefit to using python 3.10 at this point, unless you enjoy compiling things, in which case, look at numpy's source install instructions (or use their dev 3.10 wheels) before starting with `spacy`.

For the vast majority of users, it makes sense to wait for `numpy` to release 3.10 wheels first.

**Reply by gethinlw** [reply]
It's literally just so I can use the latest _python:slim_ image on a low-priority project that I barely maintain, but I agree that there are better ways to do this.

**Comment by svlandeg** [comment]
Happy to report that Python 3.10 wheels are now available with [spaCy 3.1.4](https://github.com/explosion/spaCy/releases/tag/v3.1.4) !

**Comment by GuillemGSubies** [comment]
Any updates here? I just tried to install spaCy with Python 3.10 and it didn't work


`conda create -n bo python=3.10 -y ; conda activate bo`
`pip install spacy`
`python -c "import spacy"`

The error is the following:


```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/guillem.garcia/miniconda3/envs/bo/lib/python3.10/site-packages/spacy/__init__.py", line 15, in <module>
    from .cli.info import info  # noqa: F401
  File "/home/guillem.garcia/miniconda3/envs/bo/lib/python3.10/site-packages/spacy/cli/__init__.py", line 3, in <module>
    from ._util import app, setup_cli  # noqa: F401
  File "/home/guillem.garcia/miniconda3/envs/bo/lib/python3.10/site-packages/spacy/cli/_util.py", line 9, in <module>
    import typer
  File "/home/guillem.garcia/miniconda3/envs/bo/lib/python3.10/site-packages/typer/__init__.py", line 12, in <module>
    from click.termui import get_terminal_size as get_terminal_size
ImportError: cannot import name 'get_terminal_size' from 'click.termui' (/home/guillem.garcia/miniconda3/envs/bo/lib/python3.10/site-packages/click/termui.py)
```

**Reply by adrianeboyd** [reply]
This isn't related to python 3.10. See #10564 for a workaround and for related updates in the future.
